semantic/annotate.go

```diff
@@ -0,0 +1,126 @@
+package semantic
+
+// Annotate annotates the nodes of a semantic graph with type variables
+func Annotate(node Node, venv *VariableEnvironment, tenv *TypeEnvironment) {
+	var tvar Type
+	switch n := node.(type) {
+	// Assign a new type variable for every literal
+	case *BooleanLiteral:
```

Comment:
The semantic package has helper methods for walking the graph. Are those useful here? If not why not? I'd like to try and use the walk functions for all graph traversal that way only one place needs to be updated when we add new node types.

Reply:
This was due entirely to having to pass different `VariableEnvironment`s to `Annotate` depending on the type of the expression. However I realize that `VariableEnvironment` is an unnecessary data structure, and so by getting rid of it, we can use the `semantic.Visitor` interface. This will definitely reduce the code bloat. I will update.